### PR TITLE
[do not merge] Decouple rewards from ledger

### PR DESCRIPTION
**Pull request by Ank4n**
Related to https://github.com/paritytech/substrate/issues/12847.

This PR changes the way we prevent double claim for validator rewards. We keep track of `EraRewardPoints` for each validator for current era, and in subsequent eras, when validator reward is claimed, their reward points are dropped from  `EraRewardPoints`. 

Since we store 84 eras of past reward claim history, in order to avoid doing a large migration (total nominator count * history_depth), we will do a lazy migration. After this change, no new data will be written to `StakingLedger.legacy_claimed_rewards` (renamed from `claimed_rewards`). Instead for any era going forward, we will just drop the `EraRewardPoints` for the validator as discussed above. When distributing rewards, we check `StakingLedger.legacy_claimed_rewards` to make sure if any past claim has been recorded there. 

After 84 eras. we can be fully confident that `StakingLedger.legacy_claimed_rewards` has no relevant history that we need to preserve and we can do a new migration to remove it from ledger as well as clean up the code.

## Tasks
- [x] add test to verify `claimed_rewards` is checked before paying out stakers.
- [x] stop initialising `claimed_rewards` for new stakers.
- [x] clean up exisiting tests.
- [x] re-benchmark.
- [x] Tracker issue to clean up `claimed_rewards` code, https://github.com/paritytech/substrate/issues/13034.

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2199431

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2199432

**Comment by Ank4n** [comment]
/cmd queue -c bench-bot $ pallet dev pallet_staking

**Review by gpestana — APPROVED** [review]

File: frame/staking/src/pallet/impls.rs [hunk 1]
```diff
@@ -273,6 +263,26 @@ impl<T: Config> Pallet<T> {
 		<Ledger<T>>::insert(controller, ledger);
 	}
 
+	fn get_era_reward_points(
```
Comment (kianenigma) [inline frame/staking/src/pallet/impls.rs:266]:
```suggestion
	fn temp_get_era_reward_points(
```

Comment (kianenigma) [reply frame/staking/src/pallet/impls.rs:266]: since this can be removed later, right?

**Review by kianenigma — COMMENTED** [review]
Looks mostly good, but I want to give it an in-depth review later. Also want to look into the tests again. Would be good if you revise the tests as well, as the existing ones might be mediocre at best.

**Comment by Ank4n** [comment]
While this PR will solve the existing issue of moving `rewards` away from `StakingLedger`, it will be incompatible with [multi-block rewards payout PR](https://github.com/paritytech/substrate/pull/13059) that I am working on. I am inclined to do these changes as part of the other PR to make sure it works well with multi-block rewards.

I will wait for comments/opinions but will probably close it in couple of days.

**Pull request closed without merging**